File: 1574.shortest-subarray-to-be-removed-to-make-array-sorted.cs
/*
 * @lc app=leetcode id=1574 lang=csharp
 *
 * [1574] Shortest Subarray to be Removed to Make Array Sorted
 */

// @lc code=start
public class Solution 
{
    public int bisectLeft(List<int> nums, int target)
    {
        int left = 0;
        int right = nums.Count;
        int mid;

        while (left < right)
        {
            mid = left + (right - left) / 2;
            if (nums[mid] < target)
            {
                left = mid + 1;
            }
            else
            {
                right = mid;
            }
        }
        return left;  
    }

    public int FindLengthOfShortestSubarray(int[] arr)
    {
        int n = arr.Length;
        int left;
        int right;
        int nl;
        int nr;
        int mid;
        List<int> leftSubarray = new List<int>() {arr[0]};
        List<int> rightSubarray = new List<int>();
        Stack<int> rightStack = new Stack<int>();
        rightStack.Push(arr[arr.Length - 1]);
        int p = 0;
        while (p < n - 1 && arr[p] <= arr[p + 1])
        {
            p++;
            leftSubarray.Add(arr[p]);
        }
        left = n - p - 1;
        if (left == 0)
        {
            return 0;
        }

        p = n - 1;
        while (p > 0 && arr[p - 1] <= arr[p])
        {
            p--;
            rightStack.Push(arr[p]);
        }
        right = p;

        while (rightStack.TryPop(out int item))
        {
            rightSubarray.Add(item);
        }

        nl = leftSubarray.Count;
        nr = rightSubarray.Count;
        mid = Math.Min(left, right);
        for (int i = 0; i < nl; i++)
        {
            int ind = bisectLeft(rightSubarray, leftSubarray[i]);
            mid = Math.Min(mid, n - i - 1 - nr + ind);
        }
        return mid;
    }
}
// @lc code=end

File: 1574.shortest-subarray-to-be-removed-to-make-array-sorted.py
#
# @lc app=leetcode id=1574 lang=python3
#
# [1574] Shortest Subarray to be Removed to Make Array Sorted
#

# @lc code=start
import bisect
from typing import List


class Solution:
    '''
    3 cases:     
    1. chop from start
    2. chop from end
    3. chop from the middle
    Find the option which chop the least and return    
    '''
    def findLengthOfShortestSubarray(self, arr: List[int]) -> int:
        n = len(arr)
        leftSubarray = [arr[0]]
        rightSubarray = [arr[-1]]
        
        p = 0
        while p < n - 1 and arr[p] <= arr[p + 1]:
            p += 1
            leftSubarray.append(arr[p])
        left = n - p - 1
        
        if left == 0: return 0

        p = n - 1
        while p > 0 and arr[p - 1] <= arr[p]:
            p -= 1
            rightSubarray.append(arr[p])
        right = p

        rightCorrect = []
        while rightSubarray: rightCorrect.append(rightSubarray.pop())
        
        nl, nr = len(leftSubarray), len(rightCorrect)
        mid = n
        for i in range(nl):
            ind = bisect.bisect_left(rightCorrect, leftSubarray[i])
            mid = min(mid, n - (i + 1) - (nr - ind))
        return min(mid, left, right)
            
# @lc code=end

